### ISSUE TITLE
add -nframes and -onset flags to [soundfiler]

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#2341]
as discussed in https://github.com/pure-data/pure-data/issues/1233 the idea is to add:

- A '-nframes' flag to 'read' as well, where it would define the number of samples to read from a file. If used with -resize, it would also resize the array to the number of frames (in the case there aren't as many samples as asked by -nframes, those extra points would be zeroed out). Without -resize, if -nframes reaches the size limit of the array, it gets silently truncated.

- An '-onset' flag to both 'read'/'write', which sets staring point in the array to read/write

Alongside with the other possible -onset message, -nframes and -resize would add the number of samples given by 'nframes' from the onset point and resize the table too.

**Comment by porres** [comment]
> An '-onset' flag to both 'read'/'write', which sets staring point in the array to read/write

I was quoting @Spacechild1 on this idea, but I see now that we already have a '-skip' flag for the write message that does that

**Comment by porres** [comment]
while still at it, a 'normalize' flag for the read message may also be useful and easy to implement... right? why not?

**Comment by Spacechild1** [comment]
> while still at it, a 'normalize' flag for the read message may also be useful and easy to implement... right? why not?

Totally!

**Comment by ben-wes** [comment]
after understanding that `-onset` here is related to the onset in the array (took me a bit), i tried to implement this here on my side (still needs testing). what i wonder now: how should the zero fill behave in those cases?

currently, when the soundfile is shorter than the array, the rest is filled with zeros. this should certainly stay that way. but should this behave differently with the new arguments? and should the samples before the onset be filled with zeros? i currently interpreted it on my side in a way that with these parameters set, there's no zero fill. that would allow to fill up arrays with audio snippets. is that intended?

**Comment by ben-wes** [comment]
more questions/assumptions:
* normalization should only be applied to the imported snippet and not based on values already present in the array(s), i assume?
* normalization should be applied across channels based on the max value of all channels?
* should `-nframes` with `-resize` behave similar to `-maxsize` for the case where the soundfile is smaller than maxsize? or should the array be resized to nframes samples in that way (or to onsetframes+samples in case there is an onset)?

**Comment by porres** [comment]
> normalization should only be applied to the imported snippet and not based on values already present in the array(s), i assume?

no, on the all file, this would be the same as using the normalize message to the array later

> normalization should be applied across channels based on the max value of all channels?

good question. In this case it seems like a good functionality....

but then, normalization seems like a complex issue right now, maybe put a pin on that, leave it for later.

-nframes and -onset are much clear concepts and needed anyway

**Comment by porres** [comment]
> should -nframes with -resize behave similar to -maxsize for the case where the soundfile is smaller than maxsize?

Like if it should resize the array to the number of samples in the file?

from my first comment: _If used with -resize, it would also resize the array to the number of frames (in the case there aren't as many samples as asked by -nframes, those extra points would be zeroed out)_

so, no! Just zero out the extra samples

**Comment by ben-wes** [comment]
> > normalization should only be applied to the imported snippet and not based on values already present in the array(s), i assume?

> no, on the all file, this would be the same as using the normalize message to the array later

i'm not convinced here. imho, we should only normalize the imported samples. for me, the object is mainly soundfile related and not array related. but sure, sounds good to discuss `-normalize` in another issue. 

> > should -nframes with -resize behave similar to -maxsize

> so, no! Just zero out the extra samples

ok, this seems logical as a different use case than `maxsize` then. although, the message clearly says "read number of frames (nframes) of this file". so again, zeroing out seems more array related than expected. and with `-onset`, the array should obviously be resized to nframes + onset, and not to nframes then?

**Comment by porres** [comment]
>  although, the message clearly says "read number of frames (nframes) of this file". so again, zeroing out seems more array related than expected. and with -onset, the array should obviously be resized to nframes + onset, and not to nframes then?

let's check this message in the 'write' method

-nframes <float> (maximum points to write, default from skip point to end of array)

so it writes from the skip point up to the number of samples. If the array finishes before the size of nframes, it only writes that number of samples.

For the read message it should be the same. By the way, since we have a -skip flag for write, the new flag for read should also be '-skip', so I renamed the title.

So what would it do?

It would write to the array from the skip point up to the size of 'nframes'. If the array ends before that, then it gets silently truncated. If the sound file is smaller, the extra samples are zeroed out as this is the same as what happens when you load a file into an array.

Forget about -maxsize, this is supposed to be something else, and deprecated.

Now, what happens if there is a -resize flag?

It resizes the array to meet the size of nframes (and yes, from the skip point). Again, if the sound file is smaller, the extra samples are zeroed out...

Is it clear? 

Do you think there is something inconsistent, weird, funny, not well thought?

**Comment by porres** [comment]
>  although, the message clearly says "read number of frames (nframes) of this file". so again, zeroing out seems more array related than expected. and with -onset, the array should obviously be resized to nframes + onset, and not to nframes then?

let's check this message in the 'write' method

-nframes <float> (maximum points to write, default from skip point to end of array)

so it writes from the skip point up to the number of samples. If the array finishes before the size of nframes, it only writes that number of samples.

For the read message it should be the same, but you can consider the 'onset' as the counterpart of skip... 

So what would it do?

It would write to the array from the onset point up to the size of 'nframes'. If the array ends before that, then it gets silently truncated. If the sound file is smaller, the extra samples are zeroed out as this is the same as what happens when you load a file into an array.

Forget about -maxsize, this is supposed to be something else, and deprecated.

Now, what happens if there is a -resize flag?

It resizes the array to meet the size of nframes (and yes, from the onset point). Again, if the sound file is smaller, the extra samples are zeroed out...

Is it clear? 

Do you think there is something inconsistent, weird, funny, not well thought?


You may ask, what if I don't want it to resize to the onset point + nframes? What if I want to resize to the sample size? then don't use the -nframes flag with the -resize flag :)

what if you don't want to resize? Don't use -resize ;)

makes sense? Is something funny, not well thought?

and yes... onset point plus resize resizes from the onset point

**Comment by ben-wes** [comment]
thanks for the responses.

> if there is a -resize flag [...], if the sound file is smaller, the extra samples are zeroed out...

alright with me. now special case: if there were already samples in the array in this region (that would be zeroed out in the range that exceeds the soundfile size with `-nframes <n> -resize`) - do we really really want to zero them out or keep them?

**Comment by porres** [comment]
yes, if you want to keep them, don't use the resize flag... ;)

**Comment by ben-wes** [comment]
i felt like the code got to complex with the handling of different parameter combinations (at least the way i implemented it here with these enumerated "resize_modes" and including the already present separate ascii handling) - and therefore don't want to create a PR for it that way.

but sharing this anyway for further discussion or testing if anyone wants to check it out:
* https://github.com/pure-data/pure-data/compare/develop...ben-wes:pure-data:feature/soundfiler-nframes-onset?expand=1
